src/main/java/org/pentaho/di/trans/steps/mongodbinput/MongoDbInput.java

```diff
@@ -103,20 +109,32 @@ public MongoDbInput( StepMeta stepMeta, StepDataInterface stepDataInterface, int
           }
         }
 
-        if ( meta.getOutputJson() || meta.getMongoFields() == null || meta.getMongoFields().size() == 0 ) {
+        if ( meta.getOutputJson() || meta.getMongoFields() == null || meta.getMongoFields().isEmpty() ) {
           String json = JSON.serialize( nextDoc );
           row = RowDataUtil.allocateRowData( data.outputRowMeta.size() );
-          int index = 0;
 
-          row[index++] = json;
+          if ( meta.getExecuteForEachIncomingRow() && m_currentInputRowDrivingQuery != null ) {
+            // Add the incoming columns at start and at the end of the incoming columns add json output.
+            RowMetaInterface inputRowMeta = getInputRowMeta();
+            appendTheIncomingRowsAtStart( row, inputRowMeta );
+            row [ inputRowMeta.size() ] = json;
```

Comment:
Better, but you don't need the space after row and before [ here, nor on any of the other arrays.